### ISSUE TITLE
IndexSizeError using CanvasRenderer in Firefox

### ISSUE THREAD

**Issue by cvissie** [pixijs/pixi.js#1921]
Got this error using the CanvasRenderer in Firefox.

Seems it's basically this error:
http://stackoverflow.com/questions/19338032/canvas-indexsizeerror-index-or-size-is-negative-or-greater-than-the-allowed-a

Seems like Canvas in Firefox doesn't like numbers between 0 - 1 for the width & height of draw image.

```javascript
        var w = width * resolution;
        var h = height * resolution;
        w = Math.max(1, Math.floor(w));
        h = Math.max(1, Math.floor(h));

          renderer.context.drawImage(
            texture.baseTexture.source,
            texture.crop.x * resolution,
            texture.crop.y * resolution,
            w,
            h,
            dx * renderer.resolution,
            dy * renderer.resolution,
            width * renderer.resolution,
            height * renderer.resolution
          );
```

In full, a patch for this issue as well as issue https://github.com/GoodBoyDigital/pixi.js/issues/1847

```javascript
  (function () {

    /**
     * Renders the object using the Canvas renderer
     *
     * @param renderer {CanvasRenderer} The renderer
     * @private
     */
    PIXI.Sprite.prototype._renderCanvas = function (renderer) {
      if (this.texture.crop.width <= 0 || this.texture.crop.height <= 0) {
        return;
      }

      renderer.currentBlendMode = this.blendMode;
      renderer.context.globalCompositeOperation = renderer.blendModes[renderer.currentBlendMode];

      //  Ignore null sources
      if (this.texture.valid) {
        var texture = this._texture,
          wt = this.worldTransform,
          dx,
          dy,
          width,
          height;

        renderer.context.globalAlpha = this.worldAlpha;

        // If smoothingEnabled is supported and we need to change the smoothing property for this texture
        if (renderer.smoothProperty && renderer.currentScaleMode !== texture.baseTexture.scaleMode) {
          renderer.currentScaleMode = texture.baseTexture.scaleMode;
          renderer.context[renderer.smoothProperty] = (renderer.currentScaleMode === CONST.SCALE_MODES.LINEAR);
        }

        // If the texture is trimmed we offset by the trim x/y, otherwise we use the frame dimensions

        if (texture.rotate) {

          // cheeky rotation!
          var a = wt.a;
          var b = wt.b;

          wt.a = -wt.c;
          wt.b = -wt.d;
          wt.c = a;
          wt.d = b;

          width = texture.crop.height;
          height = texture.crop.width;

          dx = (texture.trim) ? texture.trim.y - this.anchor.y * texture.trim.height : this.anchor.y * -texture._frame.height;
          dy = (texture.trim) ? texture.trim.x - this.anchor.x * texture.trim.width : this.anchor.x * -texture._frame.width;
        }
        else {
          width = texture.crop.width;
          height = texture.crop.height;

          dx = (texture.trim) ? texture.trim.x - this.anchor.x * texture.trim.width : this.anchor.x * -texture._frame.width;
          dy = (texture.trim) ? texture.trim.y - this.anchor.y * texture.trim.height : this.anchor.y * -texture._frame.height;
        }


        // Allow for pixel rounding
        if (renderer.roundPixels) {
          renderer.context.setTransform(
            wt.a,
            wt.b,
            wt.c,
            wt.d,
            (wt.tx * renderer.resolution) | 0,
            (wt.ty * renderer.resolution) | 0
          );

          dx = dx | 0;
          dy = dy | 0;
        }
        else {

          renderer.context.setTransform(
            wt.a,
            wt.b,
            wt.c,
            wt.d,
            wt.tx * renderer.resolution,
            wt.ty * renderer.resolution
          );


        }

        var resolution = texture.baseTexture.resolution;

        var w = width * resolution;
        var h = height * resolution;
        w = Math.max(1, Math.floor(w));
        h = Math.max(1, Math.floor(h));

        if (this.tint !== 0xFFFFFF) {
          if (this.cachedTint !== this.tint) {
            this.cachedTint = this.tint;

            // TODO clean up caching - how to clean up the caches?
            this.tintedTexture = CanvasTinter.getTintedTexture(this, this.tint);
          }

          renderer.context.drawImage(
            this.tintedTexture,
            0,
            0,
            w,
            h,
            dx * renderer.resolution,
            dy * renderer.resolution,
            width * renderer.resolution,
            height * renderer.resolution
          );
        }
        else {

          renderer.context.drawImage(
            texture.baseTexture.source,
            texture.crop.x * resolution,
            texture.crop.y * resolution,
            w,
            h,
            dx * renderer.resolution,
            dy * renderer.resolution,
            width * renderer.resolution,
            height * renderer.resolution
          );
        }
      }
    };


  })();
```

**Comment by cvissie** [comment]
Oh yes. It must be said, I think this happens because I'm using an odd decimal for the renderer resolution.

Renderer resolution is set at 1.8 and not 2. Optimizing for the device and the screen (mobile & desktop respectively)

**Comment by joefaron** [comment]
Having this issue too.. except your fix doesn't work with the phaser 2.4.3.  getting this:

TypeError: renderer.blendModes is undefined
renderer.context.globalCompositeOperation = renderer.blendModes[renderer.current...

**Comment by englercj** [comment]
Phaser 2 uses Pixi v2, not v3.
